lib/AST/ASTPrinter.cpp

```diff
@@ -143,14 +145,21 @@ PrintOptions PrintOptions::printSwiftInterfaceFile(bool preferTypeRepr,
       if (D->getAttrs().hasAttribute<ImplementationOnlyAttr>())
         return false;
 
-      // Skip anything that isn't 'public' or '@usableFromInline'.
+      // Skip anything that isn't 'public' or '@usableFromInline',
+      // or SPI if desired.
       if (auto *VD = dyn_cast<ValueDecl>(D)) {
         if (!isPublicOrUsableFromInline(VD)) {
           // We do want to print private stored properties, without their
           // original names present.
           if (auto *ASD = dyn_cast<AbstractStorageDecl>(VD))
             if (contributesToParentTypeStorage(ASD))
               return true;
+
+          // Always print SPI decls if `PrintSPIs`.
+          if (options.PrintSPIs &&
+              VD->getAttrs().hasAttribute<SPIAccessControlAttr>())
```

Comment:
Why is this check necessary? It should be fine if we always print SPI attributes and they will be skipped (because decls are skipped) if we print non-spi version of the interface.

Reply:
SPI decls would be skipped because there formal access is considered to be internal here, so we have to force skip them. It’s a bit backwards because of the recent change from the default internal to public, I’ll have to come back to clean this.